### PR TITLE
feat: added draft functionality for comment and responses

### PR DESCRIPTION
**Pull request by sundasnoreen12**
[INF-1444](https://2u-internal.atlassian.net/browse/INF-1444)
### Description

Learners using the discussion board have faced a frustrating issue: losing their work when switching tabs and returning to the post. This can be especially disheartening for those writing lengthy comments or responses. The issue occurs specifically while writing comments or responses on posts.

Solution 

Draft Feature: Offering a draft functionality where users can save unfinished posts and return to them later. The draft only be saved on the frontend and are lost on page reload.

#### Screenshots/sandbox (optional):
https://www.loom.com/share/fa0007a967ea406f89835409009dd3b1?sid=72cdf8c3-2e9f-4816-ba62-d7b991b5872b

#### Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Sandbox, if applicable.
* [ ] Is there adequate test coverage for your changes?

#### Post-merge Checklist

* [ ] Deploy the changes to prod after verifying on stage or ask **@openedx/edx-infinity** to do it. 
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

File: src/discussions/post-comments/data/slices.js [hunk 1]
```diff
@@ -22,6 +22,10 @@ const commentsSlice = createSlice({
     pagination: {},
     responsesPagination: {},
     sortOrder: true,
+    draft: {
+      responses: [],
+      comments: [],
+    },
```
Comment (awais-ansari) [inline src/discussions/post-comments/data/slices.js:28]: Slice:
draftResponses: {}
draftComments: {}

Example:
draftComments: {
  comment1 : {
    id: comment1,
    parentId: response1,
    content: "some content",
    postId: post1
  },
}

draftResponses: {
  response1: {
    id: response1,
    parentId: comment1,
    content: "some content",
    postId: post1
  }
}

Access Example: 
commnet1 = draftComments.?[draftId]
response1 = draftResponses.?[draftId]

Working Example

newComment = {
  id: comment1,
  parentId: response1,
  content: "some content",
  postId: post1
}

addIntoDraftComments(newComment)  // add into draftComments

addIntoDraftComments(newComment) {
  draftComments = {
    ...draftComments,
    [newComment.id]: newComment
  }
}

removeFromDraftComments(newComment) {
  const { removeComment, ...RemainingDraftCommnets} = draftComments
  return RemainingDraftCommnets;
}

**Review by awais-ansari — CHANGES_REQUESTED** [review]

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 2]
```diff
@@ -97,6 +111,34 @@ const CommentEditor = ({
...
+    setEditorContent(draftHtml);
+    initialValues.comment = draftHtml;
+  }, [parentId, threadId, id]);
+
+  const extractContent = (content) => {
+    if (typeof content === 'object') {
+      return content.target.getContent();
+    }
+    return content;
+  };
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:125]: Move it to utils.

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 3]
```diff
@@ -97,6 +111,34 @@ const CommentEditor = ({
     }
   }, [formRef]);
 
+  useEffect(() => {
+    const draftHtml = draftContent.getDraftContent(parentId, threadId, id) || rawBody;
+    setEditorContent(draftHtml);
+    initialValues.comment = draftHtml;
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:117]: `setEditorContent` will cause a re-rendering and a comment value will get assigned.

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 4]
```diff
@@ -71,6 +75,15 @@ const CommentEditor = ({
     onCloseEditor();
   }, [onCloseEditor, initialValues]);
 
+  const DeleteEditorContent = async () => {
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:78]: DeleteEditorContent => deleteEditorContent

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 5]
```diff
@@ -45,6 +47,8 @@ const CommentEditor = ({
   const userIsStaff = useSelector(selectUserIsStaff);
   const { editReasons } = useSelector(selectModerationSettings);
   const [submitting, dispatch] = useDispatchWithState();
+  const [editorContent, setEditorContent] = useState();
+  const draftContent = useDraftContent();
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:51]: const { addDraftContent, getDraftContent, removeDraftContent } = useDraftContent();

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 6]
```diff
@@ -151,7 +193,10 @@ const CommentEditor = ({
             id={editorId}
             value={values.comment}
             onEditorChange={formikCompatibleHandler(handleChange, 'comment')}
-            onBlur={formikCompatibleHandler(handleBlur, 'comment')}
+            onBlur={(content) => {
+              formikCompatibleHandler(handleChange, 'comment');
+              saveDraftContent(typeof content === 'object' ? content.target.getContent() : content);
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:198]: refactor it
saveDraftContent(typeof content === 'object' ? content.target.getContent() : content);
`saveDraftContent(content)`

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 7]
```diff
@@ -71,6 +75,15 @@ const CommentEditor = ({
...
 
+  const deleteEditorContent = async () => {
+    const { updatedResponses, updatedComments } = removeDraftContent(parentId, id, threadId);
+    if (parentId) {
+      await dispatch(setDraftComments(updatedComments));
+    } else {
+      await dispatch(setDraftResponses(updatedResponses));
+    }
+  };
+
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:86]: Apply function memorization.

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/discussions/post-comments/comments/comment/CommentEditor.jsx [hunk 8]
```diff
@@ -97,11 +111,33 @@ const CommentEditor = ({
...
+      id,
+      threadId,
+    );
+    if (parentId) {
+      await dispatch(setDraftComments(updatedComments));
+    } else {
+      await dispatch(setDraftResponses(updatedResponses));
+    }
+  };
+
```
Comment (awais-ansari) [inline src/discussions/post-comments/comments/comment/CommentEditor.jsx:134]: Please memorize this function too.

**Review by awais-ansari — APPROVED** [review]

File: src/discussions/post-comments/data/hooks.js [hunk 9]
```diff
@@ -102,3 +104,80 @@ export function useCommentsCount(postId) {
...
+      const responseObj = id ? responses[id] : getObjectByParentId(responses, threadId, false, id);
+      if (responseObj) {
+        updatedResponses = removeItem(responses, responseObj.id);
+      }
+    } else {
+      const commentObj = id ? comments[id] : getObjectByParentId(comments, parentId, true, id);
+      if (commentObj) {
+        updatedComments = removeItem(comments, commentObj.id);
+      }
+    }
```
Comment (ayesha-waris) [inline src/discussions/post-comments/data/hooks.js:173]: please consider a method for statements within if and else, the statements appear to be redundant
```
const commentObj = id ? comments[id] : getObjectByParentId(comments, parentId, true, id);
      if (commentObj) {
        updatedComments = removeItem(comments, commentObj.id);
      }
```
```
const responseObj = id ? responses[id] : getObjectByParentId(responses, threadId, false, id);
      if (responseObj) {
        updatedResponses = removeItem(responses, responseObj.id);
}

File: src/discussions/post-comments/PostCommentsView.test.jsx [hunk 10]
```diff
@@ -671,6 +671,176 @@ describe('ThreadView', () => {
...
+        fireEvent.click(cancelBtn);
+      });
+
+      addCommentBtn = screen.queryByText('Add comment');
+      await act(async () => {
+        fireEvent.click(addCommentBtn);
+      });
+
+      expect(screen.queryByText('Draft comment!')).toBeInTheDocument();
+    });
```
Comment (awais-ansari) [inline src/discussions/post-comments/PostCommentsView.test.jsx:699]: Refactoring
````
await act(async () => {
        fireEvent.click(screen.queryByText('Add comment'));
});

await act(async () => {
        fireEvent.click(screen.queryByText('Cancel'));
 });

await act(async () => {
   fireEvent.click(screen.queryByText('Add comment'));
 );

OR

const addCommentBtn = screen.queryByText('Add comment');
await act(async () => {
        fireEvent.click(addCommentBtn);
      });
      
await act(async () => {
        fireEvent.click(screen.queryByText('Cancel'));
 });
 
 await act(async () => {
        fireEvent.click(addCommentBtn);
      });

**Review by ayesha-waris — APPROVED** [review]

**Review by awais-ansari — CHANGES_REQUESTED** [review]
Please do some refactoring in test cases.

**Review by awais-ansari — APPROVED** [review]

**Pull request merged**